src/main-process/atom-application.js

```diff
@@ -1946,76 +1946,75 @@ module.exports = class AtomApplication extends EventEmitter {
   //   :path - An optional String which controls the default path to which the
   //           file dialog opens.
   promptForPathToOpen(type, { devMode, safeMode, window }, path = null) {
-    return this.promptForPath(
-      type,
-      async pathsToOpen => {
-        let targetWindow;
-
-        // Open in :window as long as no chosen paths are folders. If any chosen path is a folder, open in a
-        // new window instead.
-        if (type === 'folder') {
-          targetWindow = null;
-        } else if (type === 'file') {
+    return this.promptForPath(type, path).then(async result => {
+      const pathsToOpen = result.filePaths;
+      let targetWindow;
+
+      // Open in :window as long as no chosen paths are folders. If any chosen path is a folder, open in a
+      // new window instead.
+      if (type === 'folder') {
+        targetWindow = null;
+      } else if (type === 'file') {
+        targetWindow = window;
+      } else if (type === 'all') {
+        const areDirectories = await Promise.all(
+          pathsToOpen.map(
+            pathToOpen =>
+              new Promise(resolve => fs.isDirectory(pathToOpen, resolve))
+          )
+        );
+        if (!areDirectories.some(Boolean)) {
           targetWindow = window;
-        } else if (type === 'all') {
-          const areDirectories = await Promise.all(
-            pathsToOpen.map(
-              pathToOpen =>
-                new Promise(resolve => fs.isDirectory(pathToOpen, resolve))
-            )
-          );
-          if (!areDirectories.some(Boolean)) {
-            targetWindow = window;
-          }
         }
-
-        return this.openPaths({
-          pathsToOpen,
-          devMode,
-          safeMode,
-          window: targetWindow
-        });
-      },
-      path
-    );
-  }
-
-  promptForPath(type, callback, path) {
-    const properties = (() => {
-      switch (type) {
-        case 'file':
-          return ['openFile'];
-        case 'folder':
-          return ['openDirectory'];
-        case 'all':
-          return ['openFile', 'openDirectory'];
-        default:
-          throw new Error(`${type} is an invalid type for promptForPath`);
       }
-    })();
 
-    // Show the open dialog as child window on Windows and Linux, and as an independent dialog on macOS. This matches
-    // most native apps.
-    const parentWindow =
-      process.platform === 'darwin' ? null : BrowserWindow.getFocusedWindow();
+      return this.openPaths({
+        pathsToOpen,
+        devMode,
+        safeMode,
+        window: targetWindow
+      });
+    });
+  }
 
-    const openOptions = {
-      properties: properties.concat(['multiSelections', 'createDirectory']),
-      title: (() => {
+  promptForPath(type, path) {
```

Comment:
i m not sure how you develop a plugin for atom, but isn't there the risk that something outside atom could use this promptForPath and the signature just changed?

Reply:
If I were to try this again, I would try to copy your approach from https://github.com/atom/atom/pull/21934 / https://github.com/atom/atom/pull/21935.

That seems to be a simpler change with less side-effects.

Your approach would also probably not require me to change any of the tests! I'd like to wait and see what happens with those two Pull Requests. And I'll try to copy the final result from there.

---

If you want to go ahead and do this yourself I wouldn't stop you. I definitely wanted to report the problem, which I did in https://github.com/atom/atom/pull/21777#issuecomment-773014459. And I wanted to attempt this myself, both to learn how to use Promises, and to "get the ball rolling" and start the process, so other people would be more likely to act on it. As it turns out, I have a bit more to learn about Promises, function signatures... and so on.

Reply:
I have merged the 2 PRs #21934 and  #21935. It would be good if the approach on those 2 PR's are followed. @DeeDeeG Let me know when this PR is ready for review.

Reply:
@sadick254 your commit https://github.com/atom/atom/pull/21777/commits/e5267bcade57ffdb40e39e5cf5910eb71d43365d from the Electron 9 PR looks right.

I expect that commit works perfectly. But I'll do a quick CI run with that against `master` to confirm and see how it goes.

Reply:
Okay, I am working on this.

Two things:

1) The Promise gives an object back:

```js
{
  canceled: true|false // Boolean - whether or not the dialog was canceled.
  filePaths: [''] // String[] - An array of file paths chosen by the user. If the dialog is cancelled this will be an empty array.
  bookmarks: [''] // something specific to Mac App Store builds
}
```

So I believe the callback should be run with those as arguments like in @asturur's PR, if the callback gets called at all.

2) Might need to conditionally do `.then(callback)` only if `typeof callback === 'function'` like in @asturur's PR.